File: cpp/1.cpp
// read 1.txt line by line
// and print the line number and the line content

#include <iostream>
#include <fstream>
#include <string>
#include <map>

void part1();
void part2();

int main() {
    part1();
    part2();
    return 0;
}

void part1() {
    std::ifstream ifs("inputs/1.txt");
    std::string line;
    int sum = 0;
    while (std::getline(ifs, line)) {
        char *firstDigit = nullptr;
        char *lastDigit = nullptr;
        for (int i = 0; i < line.size(); i++) {
            if (line[i] >= '0' && line[i] <= '9') {
                if (firstDigit == nullptr) {
                    firstDigit = &line[i];
                }
                lastDigit = &line[i];
            }
        }
        if (firstDigit != nullptr && lastDigit != nullptr) {
            sum += (*firstDigit - '0') * 10 + (*lastDigit - '0');
        }
    }
    std::cout << "Part 1 Answer: " << sum << std::endl;
}

void part2() {
    // create a map of string to string
    std::map<std::string, int> numbersMapping;

    numbersMapping["zero"] = 0;
    numbersMapping["one"] = 1;
    numbersMapping["two"] = 2;
    numbersMapping["three"] = 3;
    numbersMapping["four"] = 4;
    numbersMapping["five"] = 5;
    numbersMapping["six"] = 6;
    numbersMapping["seven"] = 7;
    numbersMapping["eight"] = 8;
    numbersMapping["nine"] = 9;

    std::ifstream ifs("1.txt");
    std::string line;
    int sum = 0;
    while (std::getline(ifs, line)) {

        int firstDigit;
        int firstDigitIndex = -1;
        int lastDigit;
        int lastDigitIndex = -1;
        for (int i = 0; i < line.size(); i++) {
            if (line[i] >= '0' && line[i] <= '9') {
                if (firstDigitIndex == -1) {
                    firstDigit = line[i] - '0';
                    firstDigitIndex = i;
                }
                lastDigit = line[i] - '0';
                lastDigitIndex = i;
            }
        }

        for (const auto &pair: numbersMapping) {
            std::string::size_type n = 0;
            while (std::string::npos != (n = line.find(pair.first, n))) {
                if (n < firstDigitIndex || firstDigitIndex == -1) {
                    firstDigitIndex = n;
                    firstDigit = pair.second;
                }
                if (n > lastDigitIndex || lastDigitIndex == -1) {
                    lastDigitIndex = n;
                    lastDigit = pair.second;
                }
                n = n + pair.first.size();
            } 
        }

        sum += (firstDigit * 10) + lastDigit;
    }
    std::cout << "Part 2 Answer: " << sum << std::endl;
}

File: rust/src/1.rs
fn part1(input: &str) -> u32 {
    input
        .lines()
        .map(|line| {
            line.chars()
                .filter_map(|c| c.to_digit(10))
        })
        .map(|mut iter| {
            let first = iter.next().expect("Should contain atleast one number");
            let last = iter.last();

            match last {
                Some(last) => 10*first + last,
                None => 10*first + first
            }
        })
        .sum()
}

fn part2(input: &str) -> u32 {
    input
        .lines()
        .map(|line| {
            let line = line
                .replace("zero", "z0o")
                .replace("one", "o1e")
                .replace("two", "t2o")
                .replace("three", "t3e")
                .replace("four", "f4r")
                .replace("five", "f5e")
                .replace("six", "s6x")
                .replace("seven", "s7n")
                .replace("eight", "e8t")
                .replace("nine", "n9e");

            let mut iter = line
                .chars()
                .filter_map(|c| c.to_digit(10));

            
            let first = iter.next().expect("Should contain atleast one number");
            let last = iter.last();

            match last {
                Some(last) => 10*first + last,
                None => 10*first + first
            }
        })
        .sum()
}

fn main() {
    let input = std::fs::read_to_string("inputs/1.txt").unwrap();
    println!("Part 1 Answer: {}", part1(&input));
    println!("Part 2 Answer: {}", part2(&input));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        let input = "1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet";
        assert!(part1(input) == 142);
    }

    #[test]
    fn test_part2() {
        let input = "two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen";
        assert!(part2(input) == 281);
    }
  
}